lib/usb/usb_sam_udp.c

```diff
@@ -0,0 +1,313 @@
+	}
+}
+
+static uint8_t udp_ep_stall_get(usbd_device *dev, uint8_t addr)
+{
+	(void)dev;
+	addr &= 0x7F;
+	if (addr >= UDP_EP_COUNT) {
+		return 0;
+	}
+	return ((UDP_CSR(addr) & UDP_CSR_FORCESTALL) == UDP_CSR_FORCESTALL);
+}
+
+static void udp_ep_nak_set(usbd_device *dev, uint8_t addr, uint8_t nak)
+{
+	(void)dev;
+	/* It does not make sence to force NAK on IN endpoints. */
+	if (addr & 0x80) {
+		return;
+	}
+	if (addr >= UDP_EP_COUNT) {
+		return;
+	}
+	/* Disable the endpoint to force a NAK. */
+	if (nak) {
+		udp_csr_clear(addr, UDP_CSR_EPEDS);
+	} else {
+		udp_csr_set(addr, UDP_CSR_EPEDS);
+	}
+}
+
+static uint16_t udp_ep_write_packet(usbd_device *dev, uint8_t addr,
+					 const void *buf, uint16_t len)
+{
+	(void)dev;
+	const uint8_t *p = buf;
+	const uint8_t *end = p + len;
+	addr &= 0x7F;
+	if (addr >= UDP_EP_COUNT) {
+		return 0;
+	}
+	
+	/* TODO: This assumes a non ping-pong endpoint, we could improve throughput
+	 * on ping-pong endpoints by skipping this check to fill the 2nd FIFO, but
+	 * then we need to add some bookkeeping to know if the 2nd FIFO is already
+	 * in use or not.
+	 */
+	if ((UDP_CSR(addr) & UDP_CSR_TXPKTRDY) == UDP_CSR_TXPKTRDY) {
+		return 0;
+	}
+	
+	while (p < end) UDP_FDR(addr) = *p++;
+	udp_csr_set(addr, UDP_CSR_TXPKTRDY);
+	return len;
+}
+
+static uint16_t udp_ep_read_packet(usbd_device *dev, uint8_t addr,
+					 void *buf, uint16_t len)
+{
+	(void)dev;
+	uint32_t csr, rxbit;
+	uint8_t *p = buf;
+	uint8_t *end;
+	if (addr >= UDP_EP_COUNT) {
+		return 0;
+	}
+	
+	/* TODO: What happens when both FIFOs of a ping-pong endpoint are filled? */
+	csr = UDP_CSR(addr);
+	rxbit = csr & (UDP_CSR_RX_DATA_BK0 | UDP_CSR_RX_DATA_BK1 | UDP_CSR_RXSETUP);
+	if (rxbit == 0) {
+		return 0;
+	}
+
+	len = MIN(UDP_CSR_GET_RXBYTECNT(csr), len);
+	end = p + len;
+	while (p < end) *p++ = UDP_FDR(addr);
+	if ((csr & UDP_CSR_RXSETUP) && (*(uint8_t*)buf & 0x80)) {
+		udp_csr_set(addr, UDP_CSR_DIR);
+	}
+	udp_csr_clear(addr, rxbit);
+	return len;
+}
+
+static void udp_poll_ep(usbd_device *dev, uint8_t ep)
+{
+	uint32_t csr = UDP_CSR(ep);
+	
+	if (csr & UDP_CSR_RXSETUP) {
+		if (dev->user_callback_ctr[ep][USB_TRANSACTION_SETUP]) {
+			dev->user_callback_ctr[ep][USB_TRANSACTION_SETUP](dev, ep);
+		} else {
+			udp_csr_clear(ep, UDP_CSR_RXSETUP);
+		}
+	}
+	if (csr & (UDP_CSR_RX_DATA_BK0 | UDP_CSR_RX_DATA_BK1)) {
+		if (dev->user_callback_ctr[ep][USB_TRANSACTION_OUT]) {
+			dev->user_callback_ctr[ep][USB_TRANSACTION_OUT](dev, ep);
+		} else {
+			udp_csr_clear(ep, UDP_CSR_RX_DATA_BK0 | UDP_CSR_RX_DATA_BK1);
+		}
+	}
+	if (csr & UDP_CSR_TXCOMP) {
+		if (dev->user_callback_ctr[ep][USB_TRANSACTION_IN]) {
+			dev->user_callback_ctr[ep][USB_TRANSACTION_IN](dev, ep);
+		}
+		udp_csr_clear(ep, UDP_CSR_TXCOMP);
+	}
+	if (csr & UDP_CSR_STALLSENT) {
+		UDP_CSR(ep) = (csr | UDP_CSR_WRITE_NOP) & ~UDP_CSR_STALLSENT;
+	}
+}
+
+static void udp_poll(usbd_device *dev)
+{
+	int i;
+
+	if (UDP_ISR & UDP_INT_ENDBUSRES) {
+		dev->pm_top = 0x40; /* ??? */
```

Comment:
wat?  if you don't know, how is anyone else supposed to know?  More info here please :)